interfaces/acados_template/acados_template/mpc_utils.py

```diff
@@ -28,37 +28,355 @@
 # ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 # POSSIBILITY OF SUCH DAMAGE.;
 #
-
+from array import array
 from copy import deepcopy
-from typing import Tuple
+from typing import Tuple, Optional
 import casadi as ca
 import numpy as np
 
+from . import AcadosOcpIterate, AcadosOcpSolver
 from .acados_model import AcadosModel
 from .acados_ocp import AcadosOcp, AcadosOcpConstraints
-from .utils import casadi_length, is_empty
+from .utils import casadi_length, is_empty, array_to_float
 
 
-def create_model_with_cost_state(ocp: AcadosOcp) -> Tuple[AcadosModel, np.ndarray]:
+class AcadosCostConstraintEvaluator:
+    """
+    Limitation: values of numerical properties, such as bound values, W, zl, zu, Zu, Zl, yref,
+    etc. are taken from original AcadosOcp;
+    If they are changed in the solver, this is not taken into account in the evaluator.
     """
-    Creates a new AcadosModel with an extra state `cost_state`,
-    which has the dynamics of the cost function and slack penalties corresponding to the intermediate shooting nodes.
 
-    Note: In contrast cost_discretization='INTEGRATOR', this allows to integrate also the slack penalties.
-    Since l1 slack penalties are nondifferentiable, an accurate cost integration with the model created by this function should use many integrator steps, when slack penalties are part of the OCP formulation.
+    def __init__(self, ocp: AcadosOcp, with_parametric_bounds: bool = False):
+        ocp.make_consistent()
 
-    Returns the augmented model and the parameter values of the given AcadosOcp.
-    """
+        if with_parametric_bounds:
+            raise NotImplementedError(
+                "AcadosCostConstraintEvaluatator: with_parametric_bounds not implemented.")
+            # p_bounds = ca.MX.sym('p_bounds', 2*(ocp.dims.nh+ocp.dims.nbx+dims.nbu+dims.ng+dims.nphi), 1)
 
-    model = deepcopy(ocp.model)
-    symbol = model.get_casadi_symbol()
-    cost_state = symbol("cost_state")
-    cost_state_dot = symbol("cost_state_dot")
+        self.ocp = ocp
+
+        model = ocp.model
+        constraints = ocp.constraints
+        dims = ocp.dims
+        if casadi_length(ocp.model.z) > 0:
+            raise NotImplementedError(
+                "AcadosCostConstraintEvaluatator: not implemented for models with z.")
+
+        self.__parameter_values = np.tile(ocp.parameter_values, (ocp.dims.np, ocp.dims.N))
+        self.__p_global_values = ocp.p_global_values
 
+        self.time_steps = ocp.solver_options.time_steps
+
+        # setup casadi functions for constraints and cost
+        cost_expr = get_path_cost_expression(ocp)
+        cost_expr_e = get_terminal_cost_expression(ocp)
+
+        if ocp.model.p_global is None:
```

Comment:
this if-else block should be removed (doesn't work for SX and this is taken care of by the call to `make_consistent` in the very beginning)